torch/_dynamo/guards.py

```diff
@@ -470,6 +470,21 @@ def TENSOR_MATCH(self, guard: Guard):
                 weakref.ref(value),
             )
 
+    def PROCESS_GROUP_MATCH(self, guard: Guard):
+        """ProcessGroup state match"""
+        ref = self.arg_ref(guard)
+        value = self.get(guard.name)
+        t = type(value)
+
+        code = list()
+        # type_id encompasses NCCL vs Gloo
+        code.append(f"___check_type_id({ref}, {self.id_ref(t)})")
+        # size/rank appear to be the only other critical attributes
```

Comment:
I think we also need to guard group membership in terms of the global ranks. A counter example where guarding (world_size, rank) is not robust:

```python
def main() -> None:
    dist.init_process_group("gloo")
    pg_0 = dist.new_group(ranks=[0, 1, 2, 3])
    pg_1 = dist.new_group(ranks=[0, 2, 4, 6])

    if dist.get_rank() in [0, 1, 2, 3]:
        dist.barrier(pg_0)

    # From global rank 0's perspective,
    # - dist.get_world_size(pg_0) == dist.get_world_size(pg_1) == 0
    # - dist.get_rank(pg_0) == dist.get_rank(pg_1) == 0
    #
    # However the two groups are not interchangeable. For example, the
    # following code would deadlock:
    if dist.get_rank() == 0:
        dist.barrier(pg_1)
    if dist.get_rank() in [1, 2, 3]:
        dist.barrier(pg_0)


if __name__ == "__main__":
    lc = pet.LaunchConfig(
        min_nodes=1,
        max_nodes=1,
        nproc_per_node=8,
        run_id=str(uuid.uuid4()),
        rdzv_backend="c10d",
        rdzv_endpoint="localhost:0",
        monitor_interval=1,
    )
    pet.elastic_launch(lc, entrypoint=main)()
```

@mrshenli might be able to provide info on what to guard.

Reply:
well, this doesn't totally preclude the 'decaying' approach - as long as we also include another identifier in the decay.

however, as @voz pointed out we can also consider tracing the pg as a 'global' and simply guarding that it hasn't changed.  We can combine both ID_CHECK and matching of key attributes to deal with the issue of ambiguity based on attributes alone.